packages/core/src/step-handlers.ts

```diff
@@ -31,12 +31,72 @@ export const createStepHandlers = (motia: Motia): MotiaEventManager => {
         event: subscribe,
         handlerName: step.config.name,
         handler: async (event) => {
-          const { data, traceId } = event
+          let { data, traceId } = event
           const logger = event.logger.child({ step: step.config.name })
           const tracer = event.tracer.child(step, logger)
-
           globalLogger.debug('[step handler] received event', { event: removeLogger(event), step: name })
 
+          /**
+           * ✅ Simple runtime validator — no Zod needed.
+           * Matches the event data object with config.input.properties
+           */
+          function validateEventData(data: Record<string, any>, inputSchema: any) {
+            const missingFields: string[] = []
+            const extraFields: string[] = []
+            const typeMismatches: string[] = []
+
+            // 1️⃣ Expected schema properties
+            const schemaProps = inputSchema.properties ?? {}
+            const requiredFields: string[] = inputSchema.required ?? []
+
+            // 2️⃣ Check for missing required fields
+            for (const field of requiredFields) {
+              if (!(field in data)) missingFields.push(field)
+            }
+
+            // 3️⃣ Check for extra fields not defined in schema
+            for (const field of Object.keys(data)) {
+              if (!(field in schemaProps)) extraFields.push(field)
+            }
+
+            // 4️⃣ Check for type mismatches
+            for (const [field, def] of Object.entries(schemaProps as Record<string, any>)) {
+              if (field in data) {
+                const expectedType = (def as any)?.type
+                const actualType = Array.isArray((data as any)[field]) ? 'array' : typeof (data as any)[field]
+
+                if (expectedType && expectedType !== actualType) {
+                  typeMismatches.push(`"${field}": expected ${expectedType}, got ${actualType}`)
+                }
+              }
+            }
+
+            return { missingFields, extraFields, typeMismatches }
+          }
+
+          // ✅ Runtime validation block
+          if (step.config.input) {
+            if (typeof data !== 'object' || data === null) {
+              logger.error(`❌ Validation failed for event "${step.config.name}": data is not an object`)
+              return
+            }
+
+            const { missingFields, extraFields, typeMismatches } = validateEventData(
+              data as Record<string, any>,
+              step.config.input,
+            )
+
+            if (missingFields.length || extraFields.length || typeMismatches.length) {
+              motia.printer.printEventInputValidationError(
+                { topic: event.topic },
+                { missingFields, extraFields, typeMismatches },
+              )
+
+              logger.error(`❌ Validation failed for event "${step.config.name}"`)
+              return // stop execution if invalid
```

Comment:
let's just print a warning, we shouldn't block the data to flow.

warning means: something will not work correctly, please do something before an actual error happens

it will be very helpful in cases of errors like accessing field from undefined and so, developers will quickly realize it was a configuration issue